clang/lib/Basic/Targets/SPIR.h

```diff
@@ -114,14 +127,20 @@ class LLVM_LIBRARY_VISIBILITY SPIRTargetInfo : public TargetInfo {
     return CC_SpirFunction;
   }
 
-  llvm::Optional<LangAS> getConstantAddressSpace() const override {
-    // If we assign "opencl_constant" address space the following code becomes
-    // illegal, because it can't be cast to any other address space:
-    //
-    //   const char *getLiteral() {
-    //     return "AB";
-    //   }
-    return LangAS::opencl_global;
+  void setAddressSpaceMap(bool DefaultIsGeneric) {
+    AddrSpaceMap = DefaultIsGeneric ? &SPIRDefIsGenMap : &SPIRDefIsPrivMap;
+  }
+
+  void adjust(LangOptions &Opts) override {
+    TargetInfo::adjust(Opts);
+    // FIXME: SYCL specification considers unannotated pointers and references
+    // to be pointing to the generic address space. See section 5.9.3 of
+    // SYCL 2020 specification.
```

Comment:
Are the SYCL spec maintainers aware of this implementation issue?

Reply:
This comment is super strange, AMDGPU target is doing this exact thing (default is generic or flat address space) with a view on killing the map with private as default (see AMDGPU's `adjust` comment). NVPTX doesn't because the generic and private (local in the CUDA terminology) maps to the same address space, however the semantic used for Default matches the semantic of a flat address space.

Reply:
I would be happy to avoid these SYCL customizations, but I don't think we can do better w/o breaking SPIR ABI.
For instance, SPIR-V translator relies on mapping address space `0` to `private` address space (in OpenCL terms).
Let me know if you have other ideas.

Reply:
Not sure to whom you are answering :) but

> I would be happy to avoid these SYCL customizations, but I don't think we can do better w/o breaking SPIR ABI.

That was the underlying reason of my comment: I don't think you can avoid this but this is not unprecedented anyway. Because of that, this comment is strange.

Reply:
This comment was suggested by Anastasia here: https://reviews.llvm.org/D89909#inline-946388, https://reviews.llvm.org/D89909#inline-948325.
I'm open to improvements. Do you suggest removing "FIXME:" word from the comment?

Reply:
If I'm understanding properly, it sounds like the FIXME is wrong -- the code doesn't need to be fixed, it's basically behaving as the implementation requires it, and what needs to change is the SYCL spec based on implementation feedback. So perhaps change if it from a FIXME to a NOTE and file an issue with the SYCL spec authors?

Reply:
Changed FIXME to NOTE.

I'm not sure if there is an issue with the SYCL spec. The problem we discussed https://reviews.llvm.org/D89909 is that we now have two different mappings from AST/language address space to LLVM address space. The only difference between these is how `Default` address space is mapped. I see two different points of view on how `Default` should be treated for GPU programming languages.  Anastasia thinks `Default` should be mapped to OpenCL `private` address space, but this logic doesn't work in case of SYCL and other language mode, which do not infer address space attribute at AST level. SYCL spec authors are well aware of this issue and I think SYCL 2020 spec has clarifications for this.

@Naghasan, feel free to correct me.

Reply:
>  the code doesn't need to be fixed, it's basically behaving as the implementation requires it

That's my interpretation as well. The change to `NOTE` make sense.

> Anastasia thinks Default should be mapped to OpenCL private address space

I fail to see any rational for that. SPIR is a target, it is not bound to an input language.

> I'm not sure if there is an issue with the SYCL spec.

Oh there is a lot of issues in the spec, but I don't think this is one of them. And I'm not fully convinced there is a clang design issue either: the underlying issue is we are trying to map the SYCL/CUDA/OpenCL memory model onto C/C++ which will have limitations (pointer escape for instances).